### ISSUE TITLE
Node.js version support policy?

### ISSUE THREAD

**Issue by cvrebert** [sass/node-sass#1247]
Is there a policy on what's the oldest version of Node.js that will be supported by node-sass?
e.g. "Oldest Active LTS", "Newest Active LTS", "Oldest Stable", "Oldest Maintenance", etc.
For reference, see https://github.com/nodejs/LTS , complete with snazzy chart at the bottom.

**Comment by xzyfer** [comment]
There is not. We trying to support the latest version as soon as is reasonable. We currently support all major stable versions from 0.10 to 5. Support for newer versions changes between releases it's hard to document. We're open suggestions.

**Comment by chriseppstein** [comment]
I vote that our support should mimic whatever [NAN](https://github.com/nodejs/nan) supports. The whole point of using that library is to span multiple node releases.

**Comment by xzyfer** [comment]
@chriseppstein that's generally true. The caveat being NAN support changes between releases of node-sass for example Node 5 was release when v3.4.1 was latest so we build binaries for that release. We have not and likely regenerate binaries for previous releases. This is why v2.x doesn't support iojs etc..

**Comment by xzyfer** [comment]
Back porting binaries for older release isn't sustainable and in some cases (i.e. pre NAN releases) isn't possible.

TLDR; we support 0.10, and 0.12 for node-sass > v2. Support for additional iojs and Node major versions starts from the current latest release that was available at the time of their release.

**Comment by siemiatj** [comment]
Are you sure 3.4.1 supports node 5 ?

```
>> Error: `libsass` bindings not found. Try reinstalling `node-sass`?

Running "build:dev" (build) task
Warning: Task "sass:dev" not found. Use --force to continue.

Aborted due to warnings.
$ node -v                                                                                                                               
v5.0.0
$ node-sass -v                                                                                                                         
node-sass   3.4.1   (Wrapper)   [JavaScript]
libsass     3.3.1   (Sass Compiler) [C/C++]
```

**Comment by xzyfer** [comment]
@siemiatj I'm very sure. This error is almost always user error. It happens when you install node-sass, and then change you're environment. This usually happens when you change node version without rebuilding node_modules, or you're sharing the directory with a VM and have not rebuilt node_modules in that VM. 

Each binary is specific to a version of node, you're operating system, and architecture (32 vs 64bit). Any time this changes you need to rebuild the binary for your new environment. Previous binaries will persist so you can run node-sass on multiple environments i.e. VMs or docker containers.

In this case it looks like you have two version of node-sass installed, node-sass itself, and grunt-sass which installs its own node-sass. You grunt task is using an older version of node-sass which you likely have not rebuilt the binary for after updating to Node 5. 

You can confirm this with `npm ls node-sass`. If you try to `npm rebuild` you should get an error because your older version of node-sass isn't compatible with Node 5. Make sure you're using grunt-sass@^1 and npm update.

**Comment by elektronik2k5** [comment]
I can confirm it works for me:
```console
➜  90min git:(feature/add-avn) node -v
v5.0.0
➜  90min git:(feature/add-avn) node_modules/node-sass/bin/node-sass -v
node-sass	3.4.1	(Wrapper)	[JavaScript]
libsass  	3.3.1	(Sass Compiler)	[C/C++]
```

**Comment by cvrebert** [comment]
Following NAN seems reasonable. (As would some flavor of "Active LTS" IMHO, although that'd obviously be less generous and nobody has mentioned it so far; thus I only mention this for completeness.)

Here's some strawman phrasing I came up with for a new section of the README to document the "same as NAN" policy:
> ## Node.js version compatibility
> Each release of node-sass supports the versions of Node.js that were supported by the latest stable version of [NAN](https://github.com/nodejs/nan) at the time of the relevant node-sass release, except that we reserve the right to not add support for new Node.js major versions during patch releases of node-sass.
> As of November 2015, this means that new releases of node-sass support Node.js 0.8, 0.10, 0.12, 4, and 5.

**Comment by xzyfer** [comment]
Although technically correct, my concern is that means nothing to the
layperson.

As far as support goes we've declared bankruptcy on < 3.3. This is largely
because it was before I took over and several major LibSass versions behind.

Given this I think we can be explicitly about version support for specific
versions of node sass.

**Comment by robjens** [comment]
It would be nice I guess, but indeed: tainted env after version change cut the deal for me 4.2.2 LTS, couldn't get it run on 5.1.0 yet though.

**Comment by xzyfer** [comment]
The latest version should run on Node 5. I think the most apt description of our support policy is

>Node Sass supports the all major stable Node versions available at the specific Node Sass release

**Comment by cvrebert** [comment]
Can we assume per SemVer that (in the event it were to become necessary) dropping support for an old major version of Node would necessitate a bump of node-sass's major version number?

**Comment by xzyfer** [comment]
Given that Node now has LTS support I'm happy to agree to this.

**Comment by saper** [comment]
Depend on what we mean by 'support': on a source code level we support everything that nan supports; another question is producing binaries. I somehow wonder whether and how we might drop shipping binaries for io.js, especially older releases.

**Comment by xzyfer** [comment]
We have no intention of dropping binaries in the near future. We'll continue to monitor download stats for the moment. My company for instance it stuck on iojs 2.

**Comment by marko-avlijas** [comment]
## Front end  & non-node back end developer perspective

We don't care about node, it's versions we just want everything to work.
Only reason I am aware node version change is because nodesass stops working every once in a while. They I need to reinstall nodejs which is a pain in linux.

It seems pretty stupid that node-sass should break if it's just a wrapper for libsass which is doing all the heavy lifting. Just pick a LTS release and support it for as long as possible.

**Comment by xzyfer** [comment]
@marko-avlijas please keep your comments constructive to the issue. Inflammatory language does nothing to help any body.

>Front end & non-node back end developer perspective

Myself, and likely everyone else on this issue is a "Front end & non-node back end developer". Please don't speak for us.

>It seems pretty stupid that node-sass should break if it's just a wrapper for libsass

This is in fact not stupid, but a reality of native node extension as previously explained in this issue

>Each binary is specific to a version of node, you're operating system, and architecture (32 vs 64bit). Any time this changes you need to rebuild the binary for your new environment. Previous binaries will persist so you can run node-sass on multiple environments i.e. VMs or docker containers.

>Just pick a LTS release and support it for as long as possible.

Also as previous explained in this issue (multiple times)

>We currently support all major stable versions from 0.10 to 5.

>Node Sass supports the all major stable Node versions available at the specific Node Sass release

>We have no intention of dropping binaries [for previous Node versions] in the near future.

**Comment by marko-avlijas** [comment]
My post wasn't inflammatory, it was maybe oozing frustration because I lost lot of time on reinstalling node just to make node-sass work.

It doesn't matter if you support all latest versions from 0.10 to 5 if we still have to keep reinstalling newer and newer node every 2 months (0.10.10 then 0.10.20 then 0.10.32 or from 4.0 to 4.05 to 4.10 or whatever).

Please pick a version that will not have to be reinstalled for at least a year.

And thank you for all the work into making this tool. I really mean that.

**Comment by xzyfer** [comment]
>It doesn't matter if you support all latest versions from 0.10 to 5

I think you're misunderstanding. You shouldn't need to upgrade Node to use node-sass. It works a for all Node versions, including patches, for all major stable releases.

To be 100% clear node-sass will work for all version of 0.10. **However** if you change your version of node, your operating system or architecture (32 vs 64 bit) you'll need to recompile the native extension by either reinstalling the node-sass package (just the package) or running `npm rebuild node-sass`.

**Comment by marko-avlijas** [comment]
In that case compatibility is great, but node-sass didn't work for me yesterday on linux mint 17.2 and my node version was 0.10.27 I think. I even got the message that node-sass requires 0.10.33 I think.

I tried installing with node 4.0 from nodesource and it didn't work either, but version 0.12.9 does work.

**Comment by xzyfer** [comment]
@marko-avlijas please open a different issue for your installation errors after following and recording the steps from the [troubleshooting guide](https://github.com/sass/node-sass/blob/master/TROUBLESHOOTING.md).

**Comment by mmwtsn** [comment]
Does this documentation exist anywhere? The [error message](https://github.com/sass/node-sass/blob/cf87e0bc259dfc84bb96d47fe8603674221a795d/lib/errors.js#L35) still points to `TODO URL`.

**Comment by xzyfer** [comment]
As mentioned in the thread our node version support is "from 0.10 up to
latest node version available when a node sass version was released".

Due to the nature of native extensions major node version releases will be
incompatible. Every native extension will need to release a new version
with node 6 support.

**Comment by xzyfer** [comment]
We now add a support table to each released, and the installer will link to release docs when an unsupported platform is encountered.

**Comment by xzyfer** [comment]
Added node version support to the read me

https://github.com/sass/node-sass/blob/master/README.md#supported-nodejs-versions-010-012-1-2-3-4-and-5
